### ISSUE TITLE
How to pass and get parameters within components ?

### ISSUE THREAD

**Issue by Himanshu-Gurung** [bbc/tal#460]
Greetings All,

I am facing issues regarding the parameters/arguments passing and receiving. Using the reference[http://bbc.github.io/tal/overview/components.html],I am passing argument in component push as follows;-
this.getCurrentApplication().pushComponent("maincontainer", "sampleapp/appui/components/testcomponent"{test_key : 'test_value'});
no problem in the above as I am able to push component , but not able to get the argument/parameter via ;-
this.getCurrentApplication().getCurrentArguments();
Error : function getCurrentArguments() not defined.

Please help me out with the above.

Thank you in advance.

**Comment by sanderlooijenga** [comment]
This doesn't work because the `getCurrentArguments` method is not present on the `'antie/application'` module, but on the `'antie/widgets/componentcontainer'`.

So within a component, you can retrieve the current arguments by;

```
someMethodWithinAComponent: function () {
    // The parentWidget of a component is (most of the times) a component container
    var currentArguments = this.parentWidget.getCurrentArguments();
}
```

**Comment by Himanshu-Gurung** [comment]
Thank you for the quick response @sanderlooijenga. 

could you please share the complete component code to get arguments?

**Comment by sanderlooijenga** [comment]
You can also retrieve the arguments that you send along in the parameter passed to the component lifecycle event listeners.

I suggest you'd take a look to the provided docs; https://bbc.github.io/tal/overview/components.html

As well as take a look at the TAL example app here; https://github.com/bbc/talexample

**Comment by Himanshu-Gurung** [comment]
@sanderlooijenga, I have gone through the components docx and I am using the example app to work on. I am almost done with my project just stuck with these arguments passing. 

The brief code example will be beneficial. The demo app doesn't possess arguments passing code example.

**Comment by sanderlooijenga** [comment]
Yes it does, just look at the following line (passing of arguments);

https://github.com/bbc/talexample/blob/master/static/script/appui/components/simple.js#L91

and this line (retrieving of arguments (`evt.args`));

https://github.com/bbc/talexample/blob/master/static/script/appui/components/carouselcomponent.js#L36

**Comment by Himanshu-Gurung** [comment]
Thank you for the update @sanderlooijenga. Actually, I am working with API's and passing ids while pushing a component so passing static data and getting in **onBeforeShow** will not work for me. I need to get params in **init** after document load.

Sorry for such lengthy conversation, I am new to BBC Tal.   :)

Will be waiting...

**Comment by sanderlooijenga** [comment]
The arguments will not be passed to the constructor (`init` method) of a component, that's not how TAL works. Arguments will be passed to most/all of the component lifecycle event listeners (`beforerender`, etc.) and you can then use those arguments to fetch the data you need, as explained here;

https://bbc.github.io/tal/overview/components.html#beforerender

> (Note: If you wish to bind data to any widgets within a Component, beforerender is the correct place to do it. You should also update any Labels, Images, etc. with any content that varies based on the args property of this event.)

With an example here;

https://bbc.github.io/tal/widgets/data-binding.html#data-binding-with-a-horizontalcarousel

**Comment by Himanshu-Gurung** [comment]
Thank you for the update @sanderlooijenga. How will I be able to get any params in current component. Will I be able to push component from formatter?

**Comment by bastuijnman** [comment]
1: any `antie.widgets.ComponentContainer` is able to retrieve it's current arguments, from a `antie.widgets.Component` you can do: `this.parentWidget.getCurrentArguments() ` (as was stated by @sanderlooijenga ), provided `this` is bound to the `antie.widgets.Component`

2: You should not push components on the history stack. A Formatter should return a widget that represents the data that you put in. If you want to make the item actionable: have it fire an event on the item when you want to perform the action (so usually a `select` action). Add a listener in your component. Inside that listener verify that the item you've received is one of your actionable items and perform your action.

Example pseudo-code

```
// Formatter
define('formatter', [
    // Dependencies
], function (
    // Dependencies
) {
    return Formatter.extend({

        format: function (item) {
            var button = new button();
            button.setDataItem(item);
            return button;
        }

    });
});

define('component', [
    // Dependencies
], function (
    // Dependencies
) {
    return Component.extend({
        
        init: function () {
            this.addEventListener('select', this._onSelect.bind(this));
        },

        _onSelect: function (evnt) {
            var item = evnt.target.getDataItem();

            /*
             * Check for a data class, or if you're using an object for 
             * some property that you expect for the object you're listening for.
             */
            if (item instanceof YourClass) {
                // Do stuff
            }
        }

    })
});
```

**Comment by Himanshu-Gurung** [comment]
Thank you @bastuijnman for the update. The second one will really help.
But still facing issue with the argument passing and retrieving, please rectify my code if possible.Getting Error as **Can not read property 'getCurrentArguments' of null**.

Push component:- this.getCurrentApplication().pushComponent("maincontainer", "sampleapp/appui/components/newcomponent",{test : 'testing'});

New component:-
/**
 * @fileOverview <DESCRIPTION>
 */
require.def("sampleapp/appui/components/newcomponent",
    [
        "antie/widgets/component",
        "antie/widgets/componentcontainer"
    ],
    function (Component , ComponentContainer) {
       return Component.extend({
            init: function init () {
                var self = this;

                init.base.call(self, "newcomponent");

                var res = this.parentWidget.getCurrentArguments();

                // Add component lifecycle event listeners
                this.addEventListener("load", function(ev) { self._onLoad(ev); });
                this.addEventListener("beforerender", function(ev) { self._onBeforeRender(ev); });
                this.addEventListener("beforeshow", function(ev) { self._onBeforeShow(ev); });
                this.addEventListener("aftershow", function(ev) { self._onAfterShow(ev); });
                this.addEventListener("beforehide", function(ev) { self._onBeforeHide(ev); });
                this.addEventListener("afterhide", function(ev) { self._onAfterHide(ev); });
            },
            _onLoad: function(ev) {
                // Called when component is first loaded.
            },
            _onBeforeRender: function(ev) {
                // Called before a component is rendered.
                // This is the best place to set data-specific content.
            },
            _onBeforeShow: function(ev) {
                // Called after the component is rendered but before it is visible.
            },
            _onAfterShow: function(ev) {
                // Called after the component is visible.
            },
            _onBeforeHide: function(ev) {
                // Called before the component is hidden.
            },
            _onAfterHide: function(ev) {
                // Called after the component is hidden.
            }
        });
    }
);

**Comment by bastuijnman** [comment]
`this.parentWidget.getCurrentArguments();` when calling this in the init function there are two things happening:

* The component has not rendered yet, and this is not yet assigned a `parentWidget`
* Arguments are sent in lifecycle events, so not in the `init` function, these should be handled in `_onBeforeShow` or `_onBeforeRender`.

```
_onBeforeShow: function (ev) {
    var args = ev.args;

    // args.test === 'testing'
}
```

**Comment by ignaciolg** [comment]
Hi Himanshu,

If you are used to work with yeoman you can take a look into this
generator-tal:
https://www.npmjs.com/package/generator-tal

It will deploy a basic scaffolding with some helpful examples of the
component lifecycle and how to pass arguments between them on it (from
localVideoList.js to media.js if I'm not wrong).

Is not compatible with the latest version of TAL/BBC, but if you think that
is usefull I will update it ;D

BR


Ignacio López Gómez.

**Comment by Himanshu-Gurung** [comment]
Hello @ignaciolg ,

Thank you for the update.I will try out the repository you mentioned. Actually I am stuck with the argumenting getting part. I saw in the **tal-example** demo where arguments were passed for carousel, and how the carouselcomponent was managing those things.

But what I have managed all the listings as from API and want to push same component with different arguments[lets say ID], and according to that argument I will again get results from the API to display records.

Right I have managed all these sections, just stuck in arguments management.

Will be waiting ...

**Comment by ignaciolg** [comment]
Hi Himanshu,

I think that this answer is almost the same that the Sanders one, but here
we go.

First you need to create a widget and bind a data object to it.
(formatters/tmdb.js)

> var item, button; // not in the file ;P
> // this info comes from the tmdb.js API
> item = {
> id: 'uuid',
> poster_path: 'urlToAnImage',
> title: 'title',
> foo: 'bar'
> };
>
> button = new Button("#video-" + item.id);
> button.appendChildWidget(new Image("#img-"+item.id, util.buildImgUrl(item.
> poster_path), imgOpts));
> button.appendChildWidget(new Label(item.title));
> button.setDataItem(item);
>

Then you need to add a event listener to the widget. Inside the callback
function you will be able to retrieve the data from the widget
(components/videoList.js):

> this._carousel.addEventListener('select', this._onCarouselSelectBound);
>

At this point you will be able to use that data objet to do any action,
like an API request or pass it to a new component as arguments. In the next
example  (components/videoList.js):

> _onCarouselSelect: function (evt) {
> this.parentWidget.pushComponent('components/detail',evt.target.getDataItem
> ());
> }
>

Finally you will be able to get the dataItem in the new components using
the components lifecycle events (components/detail.js)

> // constructor, set all the lifecycle events listeners. for this example
> take a look into the beforerender one
> init: function () {
> this._super('detail');
>
> this._onLoad && this.addEventListener('load', this._onLoad.bind(this));
> this._onBeforeRender && this.addEventListener('beforerender', this.
> _onBeforeRender.bind(this));
> this._onBeforeShow && this.addEventListener('beforeshow', this.
> _onBeforeShow.bind(this));
> this._onAfterShow && this.addEventListener('aftershow', this._onAfterShow.
> bind(this));
> this._onBeforeHide && this.addEventListener('beforehide', this.
> _onBeforeHide.bind(this));
> this._onAfterhide && this.addEventListener('afterhide', this._onAfterHide.
> bind(this));
>
> this._onPlaySelectedBound = this._onPlaySelected.bind(this);
> },
>

// here you will be able to access to evt.args, that contains the second
> parameters of the pushComponent method
> _onBeforeRender: function (evt) {
> var args = evt.args || {},
> poster_path = args.poster_path || '',
> backdrop_path = args.backdrop_path || '',
> original_title = args.original_title || '',
> overview = args.overview || '';
>
> evt.stopPropagation();
> this._poster.setSrc(util.buildImgUrl(poster_path));
> this._cover.setSrc(util.buildImgUrl(backdrop_path,1280))
> this._title.setText(original_title);
> this._description.setText(overview);
>
> this._playButton.addEventListener('select', this._onPlaySelectedBound);
> }
>


To sum up you need to:

   1. Create a widget
   2. set the data you want to it with the *setDataItem* method
   3. Add an event listener
   4. Retrieve in the callback the data with the *getDataItem* method over
   the event target
   5. (optional) Use that data as arguments in a new component.


PD. All the code is from the generator-tal example


Ignacio López Gómez.

**Comment by subsidel** [comment]
Hi @Himanshu-Gurung - I'm not sure why this issue was reopened, is there anything further you need with it?
